torch/distributed/pipelining/schedules.py

```diff
@@ -1965,100 +1965,116 @@ def _assert_unsharded(stage_idx: int):
                         )
                         stage.submod.reshard()  # type: ignore[operator]
                 elif comp_type == FORWARD:
-                    if stage_uses_fsdp:
-                        _assert_unsharded(stage_idx)
-
-                    if (
-                        not stage.is_first
-                        # no recv op expected for V-schedule special case (see [Note: V-schedule special case])
-                        and not is_prev_stage_on_this_rank
-                    ):
-                        assert (
-                            stage_idx,
-                            mb_index,
-                        ) in fwd_recv_ops, f"Computing {action=} before receiving input"
-                        _wait_batch_p2p(fwd_recv_ops.pop((stage_idx, mb_index)))
-
-                    output = stage.forward_one_chunk(
-                        mb_index, arg_mbs[mb_index], kwarg_mbs[mb_index]
-                    )
-                    self._maybe_compute_loss(stage, output, target_mbs, mb_index)
+                    with record_function(f"PP::Stage{stage_idx}-Forward{mb_index}"):
```

Comment:
One way to do this with less lines of change would be to put the whole if-else chain under one big record_function block, and have a (functions.cached) helper that generates the string for it, based on the actual action string.